### PR TITLE
Seq2Seq Models Pretraining

### LINKED ISSUE #7: BART seq2seq pre-training for sentence classification and NMT
## Overview

Our current approach is based on RoBERTa model pretrained on our Python/AST code dataset which is not enough for NLG tasks such as neural machine translation or summarization. The future product should have a way to translate code changes into a short description in natural language. This feature has a seq2seq nature which means the need of such models in our research toolkit. 

mBART ([Multilingual Denoising Pre-training for Neural Machine Translation, 2001.08210](https://arxiv.org/pdf/2001.08210.pdf)) looks promising. This model is available via **🤗/Transformers** library and can be modified for an NMT pipeline with **PyTorchLightning**. 

## Roadmap

- [x] Prepare the pretraining tools & data
- [x] Wrap <s>mBART</s> BART around a seq2seq pipeline
- [x] Setup metrics for seq2seq models
- [x] Pretrain a model for sentence-classification task
- [ ] <s>Prepare / fine-tune a model for an NMT task</s>

### PR DESCRIPTION
**Pull request by mozharovsky**
# Quick Summary
* Added multiple tasks for seq2seq models pretraining
* Added translation task binarization support
* Added translation dataset for training translation-like models
* Added datasets script for reading datasets in jsonlines format
* Added support for the new pretrained tokenizers format
* Added support for setting model_max_length explicitly for new tokenizers while training
* Fixed step checkpointing scalar metrics logging
* Major modules refactoring for task compatibility
* Major step checkpoint callback refactoring for avoiding registering buffers in modules
* Major Register refactoring to use generics explicitly with strict type checking

... many more bullets, see changelog file.

# Bottomline

This PR allows to train BART, NMT and Whole Word Masking models.

### CODE REVIEW

**Comment by mozharovsky** [comment]
My todo list to merge this PR:
- [ ] <s>Merge tokenizers with their associated trainers</s>
- [x] Fix docs for training tokenizers, dataset preprocessing and training models
- [x] Write a summary of changes in PR body
- [ ] Generate changelog
